include/result.hpp

```diff
@@ -4716,28 +4728,64 @@ template <typename T, typename E>
 template <typename String, typename>
 inline RESULT_CPP14_CONSTEXPR
 auto RESULT_NS_IMPL::result<T,E>::expect(String&& message)
-  const & -> void
+  & -> typename std::add_lvalue_reference<T>::type
 {
-  if (has_error()) {
-    detail::throw_bad_result_access_message(
-      detail::forward<String>(message),
-      m_storage.storage.m_error
+  return (has_value() ||
+          (detail::throw_bad_result_access_message(
+                  detail::forward<String>(message),
+                  m_storage.storage.m_error
+          ), true),
+          m_storage.storage.m_value
+  );
+}
+
+template <typename T, typename E>
+template <typename String, typename>
+inline RESULT_CPP14_CONSTEXPR
+auto RESULT_NS_IMPL::result<T,E>::expect(String&& message)
+  && -> typename std::add_rvalue_reference<T>::type
+{
+  using reference = typename std::add_rvalue_reference<T>::type;
+
+  return (has_value() ||
+          (detail::throw_bad_result_access_message(
+                  detail::forward<String>(message),
+                  static_cast<E&&>(m_storage.storage.m_error)
+          ), true),
+          static_cast<reference>(m_storage.storage.m_value)
+  );
+}
+
+template <typename T, typename E>
+template <typename String, typename>
+inline RESULT_CPP14_CONSTEXPR
+auto RESULT_NS_IMPL::result<T,E>::expect(String&& message)
+  const & -> typename std::add_lvalue_reference<typename std::add_const<T>::type>::type
+{
+    return (has_value() ||
+            (detail::throw_bad_result_access_message(
+                    detail::forward<String>(message),
+                    m_storage.storage.m_error
+            ), true),
+            m_storage.storage.m_value
```

Comment:
Nice -- since you've updated the `expect` functions to be just a single `return` statement, I'm 99% sure that the `const`-qualified functions can be just marked `constexpr` since it'll always be `constexpr` in C++11.

I wouldn't recommend making any changes for this though; CI is currently failing because it's outdated, and so I'd rather get that fixed and verified before making the change.

Reply:
All methods are currently marked with `RESULT_CPP14_CONSTEXPR` macro, which expands to `constexpr` in the relevant cases. I think no further changes are needed, what you think?

Reply:
For now this is fine; but `RESULT_CPP14_CONSTEXPR` only expands to `constexpr` if compiling with `-std=c++14` or greater, since C++14 introduced extended `constexpr` support which allows for `constexpr` to be used on more than just single `return` statements and in non-`const`-member functions.

But I'd rather this change be separate after CI is fixed, since some older compilers have conflicting opinions on what is valid C++ and what is not, and `return` statements leveraging the comma operator are unfortunately one of those.